File: 2583/solution.py
"""
https://leetcode.com/problems/kth-largest-sum-in-a-binary-tree/description/
2583. Kth Largest Sum in a Binary Tree

You are given the root of a binary tree and a positive integer k.
The level sum in the tree is the sum of the values of the nodes that are on the same level.
Return the kth largest level sum in the tree (not necessarily distinct). If there are fewer than k levels in the tree, return -1.
Note that two nodes are on the same level if they have the same distance from the root.
"""
from typing import Optional, Deque, List
from heapq import heappush, heappop
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        queue: Deque[TreeNode] = deque()
        heap: List[int] = []

        if root is None:
            return -1

        queue.append(root)
        while queue:
            total = 0
            for _ in range(len(queue)):
                node = queue.popleft()
                total += node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            heappush(heap, total)
            if k < len(heap):
                heappop(heap)

        return heappop(heap) if len(heap) == k else -1

File: 2583/solution.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution {}

use std::cell::RefCell;
use std::cmp::Reverse;
use std::collections::{BinaryHeap, VecDeque};
use std::rc::Rc;
impl Solution {
    pub fn kth_largest_level_sum(root: Option<Rc<RefCell<TreeNode>>>, k: i32) -> i64 {
        let mut queue: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        let mut heap: BinaryHeap<Reverse<i64>> = BinaryHeap::new();
        let k = k as usize;

        if let Some(node) = root {
            queue.push_back(Rc::clone(&node));
        } else {
            return -1;
        }

        let mut node_count = queue.len();
        let mut total: i64 = 0;
        while let Some(node) = queue.pop_front() {
            node_count -= 1;
            let node = node.borrow();
            total += node.val as i64;
            if let Some(node) = &node.left {
                queue.push_back(Rc::clone(node));
            }
            if let Some(node) = &node.right {
                queue.push_back(Rc::clone(node));
            }
            if node_count == 0 {
                heap.push(Reverse(total));
                total = 0;
                node_count = queue.len();
                if k < heap.len() {
                    heap.pop();
                }
            }
        }

        if heap.len() == k {
            heap.pop().unwrap().0
        } else {
            -1
        }
    }
}
